elyra/pipeline/pipeline_definition.py

```diff
@@ -0,0 +1,281 @@
+        self._node = node
+
+    @property
+    def id(self) -> str:
+        return self._node.get('id')
+
+    @property
+    def pipeline(self) -> Dict:
+        return self._pipeline_definition
+
+    @property
+    def dict(self) -> Dict:
+        return self._node
+
+    def get(self, key: str, default_value=None) -> Any:
+        self._validate()
+        return self._node['app_data'].get(key, default_value)
+
+    def set(self, key: str, value: Any):
+        if not key:
+            raise ValueError("Key is required")
+
+        if not value:
+            raise ValueError("Value is required")
+
+        self._node['app_data'][key] = value
+
+    def validate(self) -> list:
+        pass
+
+    def is_valid(self) -> bool:
+        return len(self.validate()) == 0
+
+
+class Pipeline(NodeBase):
+
+    def __init__(self, pipeline_definition: Dict, node: Dict):
+        super().__init__(pipeline_definition, node)
+
+    @property
+    def version(self) -> int:
+        return int(self._node['app_data'].get('version'))
+
+    @property
+    def runtime(self) -> str:
+        return self._node['app_data'].get('runtime') or 'local'
+
+    @property
+    def type(self):
+        type_description_to_type = {'Kubeflow Pipelines': 'kfp',
+                                    'Apache Airflow': 'airflow',
+                                    'Generic': 'generic'}
+
+        pipeline_type_description = self._node['app_data']['properties'].get('runtime')
+        if pipeline_type_description not in type_description_to_type.keys():
+            raise ValueError(f'Unsupported pipeline runtime: {pipeline_type_description}')
+        return type_description_to_type[pipeline_type_description]
+
+    @property
+    def nodes(self) -> list:
+        if 'nodes' not in self._node:
+            raise ValueError("Pipeline is missing 'nodes' field.")
+        elif len(self._node['nodes']) == 0:
+            raise ValueError("Pipeline has zero length 'nodes' field.")
+
+        nodes: list = list()
+        for node in self._node['nodes']:
+            nodes.append(Node(self._pipeline_definition, node))
+
+        return nodes
+
+    def get_property(self, key: str, default_value=None) -> Any:
+        self._validate()
+        return self._node['app_data']['properties'].get(key, default_value)
+
+    def set_property(self, key: str, value: Any):
+        if not key:
+            raise ValueError("Key is required")
+
+        if not value:
+            raise ValueError("Value is required")
+
+        self._node['app_data']['properties'][key] = value
+
+    def validate(self) -> list:
+        validation_issues = list()
+        if 'properties' not in self._node['app_data']:
+            validation_issues.append("Node is missing 'properties' field.")
+        elif len(self._node['app_data']['properties']) == 0:
+            validation_issues.append("Pipeline has zero length 'properties' field.")
+
+        return validation_issues
```

Comment:
Seems like a `Pipeline`'s validation should call `validate()` on each of its `Node` instances.

Again, we should consider only performing validation when "dirty".